server/redeemer.go

```diff
@@ -0,0 +1,383 @@
+func NewRedeemer(recipient ethcommon.Address, eth eth.LivepeerEthClient, sm *pm.LocalSenderMonitor) (*Redeemer, error) {
+
+	if recipient == (ethcommon.Address{}) {
+		return nil, fmt.Errorf("must provide a recipient")
+	}
+
+	if eth == nil {
+		return nil, fmt.Errorf("must provide a LivepeerEthClient")
+	}
+
+	if sm == nil {
+		return nil, fmt.Errorf("must provide a SenderMonitor")
+	}
+
+	return &Redeemer{
+		recipient: recipient,
+		eth:       eth,
+		sm:        sm,
+		quit:      make(chan struct{}),
+	}, nil
+}
+
+// Start starts a Redeemer server
+func (r *Redeemer) Start(host *url.URL) error {
+	listener, err := gonet.Listen("tcp", host.String())
+	if err != nil {
+		return err
+	}
+	defer listener.Close()
+	if err != nil {
+		return err
+	}
+	// slice of gRPC options
+	// Here we can configure things like TLS
+	opts := []grpc.ServerOption{}
+	// var s *grpc.Server
+	s := grpc.NewServer(opts...)
+	r.server = s
+
+	net.RegisterTicketRedeemerServer(s, r)
+
+	return s.Serve(listener)
+}
+
+// Stop stops the Redeemer server
+func (r *Redeemer) Stop() {
+	close(r.quit)
+	r.server.Stop()
+}
+
+// QueueTicket adds a ticket to the ticket queue
+func (r *Redeemer) QueueTicket(ctx context.Context, ticket *net.Ticket) (*net.QueueTicketRes, error) {
+	t := pmTicket(ticket)
+	if r.recipient != t.Recipient {
+		return nil, status.Error(codes.PermissionDenied, fmt.Sprintf("invalid ticket recipient 0x%x, expected %v", ticket.TicketParams.Recipient, r.recipient.Hex()))
+	}
+	if err := r.sm.QueueTicket(t); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	glog.Infof("ticket queued sender=0x%x", ticket.Sender)
+
+	return &net.QueueTicketRes{}, nil
+}
+
+// MaxFloat is a unary RPC method to request the max float value for a sender
+func (r *Redeemer) MaxFloat(ctx context.Context, req *net.MaxFloatReq) (*net.MaxFloatUpdate, error) {
+	maxFloat, err := r.sm.MaxFloat(ethcommon.BytesToAddress(req.Sender))
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &net.MaxFloatUpdate{MaxFloat: maxFloat.Bytes()}, nil
+}
+
+// MonitorMaxFloat starts a server-side stream to the client to send max float updates for sender specified in the request
+func (r *Redeemer) MonitorMaxFloat(req *net.MaxFloatReq, stream net.TicketRedeemer_MonitorMaxFloatServer) error {
+	sender := ethcommon.BytesToAddress(req.Sender)
+
+	sink := make(chan struct{}, 10)
+	sub := r.sm.SubscribeMaxFloatChange(sender, sink)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case <-r.quit:
+			return nil
+		case err := <-sub.Err():
+			if err == nil {
+				return status.Error(codes.Canceled, "subscription closed")
+			}
+			return status.Error(codes.Canceled, err.Error())
+		case <-sink:
+			maxFloat, err := r.sm.MaxFloat(sender)
+			if err != nil {
+				glog.Error(err)
+			}
+			if err := stream.Send(&net.MaxFloatUpdate{MaxFloat: maxFloat.Bytes()}); err != nil {
+				glog.Error(err)
```

Comment:
According to the [docs](https://godoc.org/google.golang.org/grpc#ServerStream), any non-EOF error will abort the stream so I'm not sure if it is safe to continue using this stream in the loop in that scenario. May be safer to just return here.

Reply:
fixed in https://github.com/livepeer/go-livepeer/pull/1513/commits/6b7ac892cd933ab613b66a1cd5ae30cd990c128e